pkg/tsdb/tempo/search_stream.go

```diff
@@ -0,0 +1,160 @@
+package tempo
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/tsdb/tempo/kinds/dataquery"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+const SearchPathPrefix = "search/"
+
+type ExtendedResponse struct {
+	*tempopb.SearchResponse
+	State dataquery.SearchStreamingState
+}
+
+type StreamSender interface {
+	SendFrame(frame *data.Frame, include data.FrameInclude) error
+	SendJSON(data []byte) error
+	SendBytes(data []byte) error
+}
+
+func (s *Service) runSearchStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender, datasource *Datasource) error {
+	response := &backend.DataResponse{}
+
+	var backendQuery *backend.DataQuery
+	err := json.Unmarshal(req.Data, &backendQuery)
+	if err != nil {
+		response.Error = fmt.Errorf("error unmarshaling backend query model: %v", err)
+		return err
+	}
+
+	var sr *tempopb.SearchRequest
+	err = json.Unmarshal(req.Data, &sr)
+	if err != nil {
+		response.Error = fmt.Errorf("error unmarshaling Tempo query model: %v", err)
+		return err
+	}
+
+	if sr.GetQuery() == "" {
+		return fmt.Errorf("query is empty")
+	}
+
+	sr.Start = uint32(backendQuery.TimeRange.From.Unix())
+	sr.End = uint32(backendQuery.TimeRange.To.Unix())
+
+	stream, err := datasource.StreamingClient.Search(ctx, sr)
+	if err != nil {
+		s.logger.Error("Error Search()", "err", err)
+		return err
+	}
+
+	return s.processStream(stream, sender)
+}
+
+func (s *Service) processStream(stream tempopb.StreamingQuerier_SearchClient, sender StreamSender) error {
+	var traceList []*tempopb.TraceSearchMetadata
+	var metrics *tempopb.SearchMetrics
+	for {
+		msg, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			if err := sendResponse(&ExtendedResponse{
+				State: dataquery.SearchStreamingStateDone,
+				SearchResponse: &tempopb.SearchResponse{
+					Metrics: metrics,
+					Traces:  traceList,
+				},
+			}, sender); err != nil {
+				return sendError(err, sender)
+			}
+			break
+		}
+		if err != nil {
+			s.logger.Error("Error receiving message", "err", err)
+			return sendError(err, sender)
+		}
+
+		metrics = msg.Metrics
+		traceList = append(traceList, msg.Traces...)
+		traceList = removeDuplicates(traceList)
```

Comment:
What duplicates are there? Is tempo sending the whole list of traces on each update? If so we wouldn't need to keep track of the traces so far but otherwise not sure why there should be duplicates.